File: java/poker/src/main/java/com/duchen/poker/Poker.java
package com.duchen.poker;

import com.duchen.poker.model.Player;

import java.util.Random;

public class Poker {

    public static int PLAYER_NUM = 3;

    public static void main(String[] args) {
        Poker poker = new Poker();
        poker.startNewGame();
    }

    Player[] mPlayers = new Player[PLAYER_NUM];
    CardLibrary mCardLibrary;
    int mStartIndex = 0;

    private void startNewGame() {
        mCardLibrary = new CardLibrary();
        mStartIndex = new Random().nextInt(PLAYER_NUM);

        for (int i = 0; i < PLAYER_NUM; i++) {
            mPlayers[getActualIndex(i + mStartIndex)] = new Player();
            mPlayers[getActualIndex(i + mStartIndex)].setCards(mCardLibrary.getRoundOneCard());
        }

        for (int i = 0; i < PLAYER_NUM; i++) {
            if (mPlayers[getActualIndex(i + mStartIndex)].askToBeBoss()) {
                mPlayers[getActualIndex(i + mStartIndex)].setRole(Player.Role.BOSS);
                mPlayers[getActualIndex(i + mStartIndex)].becomeBoss(mCardLibrary.getRoundTwoCard());
                mPlayers[getActualIndex(i + 1 + mStartIndex)].setRole(Player.Role.FARMER_NEXT);
                mPlayers[getActualIndex(i - 1 + mStartIndex)].setRole(Player.Role.FARMER_PRE);
                break;
            }
        }

        for (int i = 0; i < PLAYER_NUM; i++) {
            System.out.println(mPlayers[i].toString());
        }
    }

    int getActualIndex(int i) {
        if (i < 0) {
            return 0;
        }
        if (i >= PLAYER_NUM) {
            return i % PLAYER_NUM;
        }
        return i;
    }

}

File: java/poker/src/main/kotlin/poker/Poker.kt
package poker

import poker.model.Player

import java.util.Random

class Poker {

    var mPlayers = arrayOfNulls<Player>(PLAYER_NUM)
    var mCardLibrary = CardLibrary()
    var mStartIndex = 0
    var mBossIndex = 0

    fun startNewGame() {
        mCardLibrary = CardLibrary()
        mStartIndex = Random().nextInt(PLAYER_NUM)

        for (i in 0 until PLAYER_NUM) {
            mPlayers[getActualIndex(i + mStartIndex)] = Player()
        }

        for (i in 0 until 17) {
            for (j in 0 until PLAYER_NUM) {
                mPlayers[getActualIndex(j + mStartIndex)]!!.addCard(mCardLibrary.oneCard)
            }
        }

        for (i in 0 until PLAYER_NUM) {
            if (mPlayers[getActualIndex(i + mStartIndex)]!!.askToBeBoss()) {
                mBossIndex = getActualIndex(i + mStartIndex)
                mPlayers[mBossIndex]!!.role = Player.Role.BOSS
                mPlayers[mBossIndex]!!.addCards(mCardLibrary.roundTwoCard)
                mPlayers[getActualIndex(mBossIndex + 1)]!!.role = Player.Role.FARMER_NEXT
                mPlayers[getActualIndex(mBossIndex - 1)]!!.role = Player.Role.FARMER_PRE
                break
            }
        }

        for (i in 0 until PLAYER_NUM) {
            println(mPlayers[i].toString())
        }

    }

    private fun getActualIndex(i: Int): Int {
        if (i < 0) {
            return 0
        }
        return if (i >= PLAYER_NUM) {
            i % PLAYER_NUM
        } else i
    }

    companion object {
        var PLAYER_NUM = 3
    }

}
